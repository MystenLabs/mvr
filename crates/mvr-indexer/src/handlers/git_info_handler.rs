use crate::handlers::MoveObjectProcessor;
use crate::models::mainnet::sui::dynamic_field::Field;
use crate::models::{mainnet, testnet};
use anyhow::Error;
use async_trait::async_trait;
use diesel::query_dsl::methods::FilterDsl;
use diesel::upsert::excluded;
use diesel::ExpressionMethods;
use diesel_async::RunQueryDsl;
use move_types::MoveStruct;
use mvr_schema::models::GitInfo;
use serde::de::DeserializeOwned;
use std::marker::PhantomData;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_pg_db::Connection;
use sui_types::base_types::MoveObjectType;
use sui_types::full_checkpoint_content::CheckpointData;
use crate::handlers::convert_struct_tag;

pub type MainnetGitInfo = Field<u64, mainnet::mvr_metadata::git::GitInfo>;
pub type TestnetGitInfo = Field<u64, testnet::mvr_metadata::git::GitInfo>;

pub struct GitInfoHandler<T> {
    chain_id: String,
    type_: MoveObjectType,
    phantom_data: PhantomData<T>,
}

impl<T: MoveStruct> GitInfoHandler<T> {
    pub fn new(chain_id: String) -> Self {
        // Indexing dynamic field object Field<u64, [metadata_pkg_id]::git::GitInfo>
        let struct_tag = Field::<u64, MoveGitInfo>::struct_type();
        GitInfoHandler {
            chain_id,
            type_: MoveObjectType::from(convert_struct_tag(struct_tag)),
            phantom_data: Default::default(),
        }
    }
}

impl MoveObjectProcessor<MainnetGitInfo, GitInfo> for GitInfoHandler<MainnetGitInfo> {
    const PROC_NAME: &'static str = "GitInfo - Mainnet";

    fn process_move_object(
        chain_id: String,
        move_obj: MainnetGitInfo,
        obj: &Object,
    ) -> Result<GitInfo, Error> {
        let mainnet::mvr_metadata::git::GitInfo {
            repository,
            path,
            tag,
        } = move_obj.value;
        Ok(GitInfo {
            table_id: obj.owner.get_owner_address()?.to_string(),
            object_version: obj.version().value() as i64,
            version: move_obj.name as i32,
            chain_id,
            repository: Some(repository),
            path: Some(path),
            tag: Some(tag),
        })
    }
}

impl MoveObjectProcessor<TestnetGitInfo, GitInfo> for GitInfoHandler<TestnetGitInfo> {
    const PROC_NAME: &'static str = "GitInfo - Testnet";

    fn process_move_object(
        chain_id: String,
        move_obj: TestnetGitInfo,
        obj: &Object,
    ) -> Result<GitInfo, Error> {
        let testnet::mvr_metadata::git::GitInfo {
            repository,
            path,
            tag,
        } = move_obj.value;
        Ok(GitInfo {
            table_id: obj.owner.get_owner_address()?.to_string(),
            object_version: obj.version().value() as i64,
            version: move_obj.name as i32,
            chain_id,
            repository: Some(repository),
            path: Some(path),
            tag: Some(tag),
        })
    }
}

#[async_trait]
impl<T: MoveStruct + DeserializeOwned> Handler for GitInfoHandler<T>
where
    Self: MoveObjectProcessor<T, GitInfo>,
{
    async fn commit(values: &[Self::Value], conn: &mut Connection<'_>) -> anyhow::Result<usize> {
        use mvr_schema::schema::git_infos::columns::*;
        use mvr_schema::schema::git_infos::dsl::git_infos;
        Ok(diesel::insert_into(git_infos)
            .values(values)
            .on_conflict((table_id, version))
            .do_update()
            .set((
                object_version.eq(excluded(object_version)),
                repository.eq(excluded(repository)),
                path.eq(excluded(path)),
                tag.eq(excluded(tag)),
            ))
            .filter(object_version.lt(excluded(object_version)))
            .execute(conn)
            .await?)
    }
}

impl<T: MoveStruct + DeserializeOwned> Processor for GitInfoHandler<T>
where
    Self: MoveObjectProcessor<T, GitInfo>,
{
    const NAME: &'static str = Self::PROC_NAME;
    type Value = GitInfo;

    fn process(&self, checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
        checkpoint
            .transactions
            .iter()
            .try_fold(vec![], |result, tx| {
                tx.output_objects.iter().try_fold(result, |mut result, o| {
                    if matches!(o.type_(), Some(t) if t == &self.type_) {
                        if let Some(move_obj) = o.data.try_as_move() {
                            result.push(Self::process_move_object(
                                self.chain_id.clone(),
                                bcs::from_bytes(move_obj.contents())?,
                                o,
                            )?)
                        }
                    }
                    Ok(result)
                })
            })
    }
}
