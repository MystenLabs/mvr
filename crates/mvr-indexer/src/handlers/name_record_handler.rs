use crate::models::mvr_core;
use crate::models::mvr_core::name::Name;
use async_trait::async_trait;
use diesel::upsert::excluded;
use diesel::ExpressionMethods;
use diesel_async::RunQueryDsl;
use itertools::Itertools;
use mvr_schema::models::NameRecord;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_pg_db::Connection;
use sui_types::base_types::MoveObjectType;
use sui_types::dynamic_field::{DynamicFieldInfo, Field};
use sui_types::full_checkpoint_content::CheckpointData;

const TESTNET_KEY: &str = "testnet";

pub struct NameRecordHandler {
    type_: MoveObjectType,
}

impl NameRecordHandler {
    pub fn new() -> Self {
        // Indexing dynamic field object Field<[mvr_core]::name::Name, [mvr_core]::app_record::AppRecord>
        let name_type = mvr_core::name::Name::type_();
        let app_record = mvr_core::app_record::AppRecord::type_();
        let type_ = MoveObjectType::from(DynamicFieldInfo::dynamic_field_type(
            name_type.into(),
            app_record.into(),
        ));
        NameRecordHandler { type_ }
    }
}
#[async_trait]
impl Handler for NameRecordHandler {
    async fn commit(values: &[Self::Value], conn: &mut Connection<'_>) -> anyhow::Result<usize> {
        use mvr_schema::schema::name_records::dsl::name_records;
        use mvr_schema::schema::name_records::columns::*;
        Ok(diesel::insert_into(name_records)
            .values(values)
            .on_conflict(name)
            .do_update()
            .set((
                mainnet_id.eq(excluded(mainnet_id)),
                testnet_id.eq(excluded(testnet_id)),
                metadata.eq(excluded(metadata))
            ))
            .execute(conn)
            .await?)
    }
}

impl Processor for NameRecordHandler {
    const NAME: &'static str = "NameRecord";
    type Value = NameRecord;

    fn process(&self, checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
        checkpoint
            .transactions
            .iter()
            .try_fold(vec![], |result, tx| {
                tx.output_objects.iter().try_fold(result, |mut result, o| {
                    if matches!(o.type_(), Some(t) if t == &self.type_) {
                        if let Some(move_obj) = o.data.try_as_move() {
                            use crate::models::mvr_core::app_record::AppRecord as MoveAppRecord;
                            let data: Field<Name, MoveAppRecord> =
                                bcs::from_bytes(move_obj.contents())?;
                            let Name { org, app } = data.name;
                            let MoveAppRecord {
                                app_info,
                                networks,
                                metadata,
                                ..
                            } = data.value;
                            result.push(NameRecord {
                                // TODO: use correct name to string
                                name: format!("{org:?}/{}", app.join("/")),
                                mainnet_id: app_info
                                    .and_then(|info| Some(info.package_info_id?.to_string())),
                                testnet_id: networks
                                    .get(&TESTNET_KEY.into())
                                    .and_then(|info| Some(info.package_info_id?.to_string())),
                                metadata: serde_json::to_value(metadata)?,
                            })
                        }
                    }
                    Ok(result)
                })
            })
    }
}
