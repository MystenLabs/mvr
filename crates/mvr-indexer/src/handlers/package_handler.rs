use crate::{MAINNET_CHAIN_ID, TESTNET_CHAIN_ID};
use chrono::DateTime;
use diesel_async::RunQueryDsl;
use itertools::Itertools;
use move_binary_format::CompiledModule;
use mvr_schema::models::{Package, PackageDependency};
use std::collections::HashSet;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_pg_db::{Connection, Db};
use sui_types::base_types::ObjectID;
use sui_types::full_checkpoint_content::CheckpointData;
use sui_types::move_package::MovePackage;
use sui_types::object::Data;

pub struct PackageHandler<const MAINNET: bool>;

impl<const MAINNET: bool> PackageHandler<MAINNET> {
    const CHAIN_ID: &'static str = if MAINNET {
        MAINNET_CHAIN_ID
    } else {
        TESTNET_CHAIN_ID
    };
}

#[async_trait::async_trait]
impl<const MAINNET: bool> Handler for PackageHandler<MAINNET> {
    type Store = Db;

    async fn commit<'a>(
        values: &[Self::Value],
        conn: &mut Connection<'a>,
    ) -> anyhow::Result<usize> {
        use mvr_schema::schema::package_dependencies::dsl::package_dependencies;
        use mvr_schema::schema::packages::dsl::packages;

        let deps = values.iter().flat_map(|p| &p.deps).collect::<Vec<_>>();

        let insert_package = diesel::insert_into(packages)
            .values(values)
            .on_conflict_do_nothing()
            .execute(conn)
            .await?;

        let insert_deps = diesel::insert_into(package_dependencies)
            .values(deps)
            .on_conflict_do_nothing()
            .execute(conn)
            .await?;

        Ok(insert_package + insert_deps)
    }
}

impl<const MAINNET: bool> Processor for PackageHandler<MAINNET> {
    const NAME: &'static str = if MAINNET {
        "Package - Mainnet"
    } else {
        "Package - Testnet"
    };
    type Value = Package;

    fn process(&self, checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
        let timestamp =
            DateTime::from_timestamp_millis(checkpoint.checkpoint_summary.timestamp_ms as i64)
                .unwrap()
                .naive_utc();

        let mut results = vec![];
        for tx in &checkpoint.transactions {
            for o in &tx.output_objects {
                if let Data::Package(p) = &o.data {
                    let package_id = p.id().to_hex_uncompressed();
                    let deps = package_dependencies(Self::CHAIN_ID.to_string(), p)?;
                    results.push(Package {
                        package_id,
                        original_id: p.original_package_id().to_hex_uncompressed(),
                        package_version: p.version().value() as i64,
                        move_package: bcs::to_bytes(p)?,
                        chain_id: Self::CHAIN_ID.to_string(),
                        tx_hash: tx.transaction.digest().base58_encode(),
                        sender: tx.transaction.sender_address().to_string(),
                        timestamp,
                        deps,
                    })
                }
            }
        }
        Ok(results)
    }
}

pub fn package_dependencies(
    chain_id: String,
    package: &MovePackage,
) -> Result<Vec<PackageDependency>, anyhow::Error> {
    let package_id = package.id().to_hex_uncompressed();
    let mut immediate_dependencies = HashSet::new();
    for (_, bytes) in package.serialized_module_map() {
        let module = CompiledModule::deserialize_with_defaults(bytes)?;
        immediate_dependencies.extend(
            module
                .immediate_dependencies()
                .into_iter()
                .map(|dep| ObjectID::from(*dep.address())),
        );
    }
    Ok(package
        .linkage_table()
        .iter()
        .map(|(_, u)| u.upgraded_id)
        .dedup()
        .map(|dependency_package_id| PackageDependency {
            package_id: package_id.clone(),
            dependency_package_id: dependency_package_id.to_hex_uncompressed(),
            chain_id: chain_id.clone(),
            immediate_dependency: immediate_dependencies.contains(&dependency_package_id),
        })
        .collect())
}
