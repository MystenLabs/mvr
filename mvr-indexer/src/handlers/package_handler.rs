use crate::models::{Package, SuiEnv};
use diesel_async::RunQueryDsl;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_pg_db::Connection;
use sui_sdk::types::full_checkpoint_content::{CheckpointData, CheckpointTransaction};

pub struct PackageHandler {
    sui_env: SuiEnv,
}

impl PackageHandler {
    pub fn new(sui_env: SuiEnv) -> Self {
        Self { sui_env }
    }
}

#[async_trait::async_trait]
impl Handler for PackageHandler {
    async fn commit(values: &[Self::Value], conn: &mut Connection<'_>) -> anyhow::Result<usize> {
        use crate::schema::packages::dsl::packages;
        Ok(diesel::insert_into(packages)
            .values(values)
            .on_conflict_do_nothing()
            .execute(conn)
            .await?)
    }
}

impl Processor for PackageHandler {
    const NAME: &'static str = "Package";
    type Value = Package;

    fn process(&self, checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
        // todo: store sui env as well in case we need to clean up testnet data after wipe.
        checkpoint
            .transactions
            .iter()
            .try_fold(vec![], |mut results, tx| {
                if contain_package(tx) {
                    results.append(
                        &mut tx
                            .output_objects
                            .iter()
                            .map(|o| {
                                o.data
                                    .try_as_package()
                                    .iter()
                                    .map(|p| {
                                        Ok::<_, anyhow::Error>(Package {
                                            package_id: p.id().to_hex_literal(),
                                            original_id: p.original_package_id().to_hex_literal(),
                                            package_version: p.version().value() as i64,
                                            move_package: bcs::to_bytes(p)?,
                                        })
                                    })
                                    .collect::<Vec<_>>()
                            })
                            .flatten()
                            .collect::<anyhow::Result<Vec<_>>>()?,
                    )
                }
                Ok(results)
            })
    }
}

fn contain_package(tx: &CheckpointTransaction) -> bool {
    tx.output_objects.iter().any(|o| o.is_package())
}
