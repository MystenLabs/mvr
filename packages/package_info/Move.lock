# @generated by Move, please check-in and do not edit manually.

[move]
version = 2
manifest_digest = "6EAF9E5D09667BDD9425A1DA5C58682515E7136A6DC12E3F5013480775A38409"
deps_digest = "F8BBB0CCB2491CA29A3DF03D6F92277A4F3574266507ACD77214D37ECA3F3082"
dependencies = [
  { name = "Sui" },
]

[[move.package]]
name = "MoveStdlib"
source = { git = "https://github.com/MystenLabs/sui.git", rev = "framework/testnet", subdir = "crates/sui-framework/packages/move-stdlib" }

[[move.package]]
name = "Sui"
source = { git = "https://github.com/MystenLabs/sui.git", rev = "framework/testnet", subdir = "crates/sui-framework/packages/sui-framework" }

dependencies = [
  { name = "MoveStdlib" },
]

[move.toolchain-version]
compiler-version = "1.24.0"
edition = "2024.beta"
flavor = "sui"

[env]

[env.testnet]
chain-id = "4c78adac"
original-published-id = "0x08ad4adcb87d72b0c1626ee1c50cd1c0a9642c703b91df49399f44880d774841"
latest-published-id = "0x08ad4adcb87d72b0c1626ee1c50cd1c0a9642c703b91df49399f44880d774841"
published-version = "1"

[env.mainnet]
chain-id = "35834a8a"
original-published-id = "0x15f29668bc8a8168975f6de8da926ea398619cfff5894bb71d31a777fa7ea18e"
latest-published-id = "0x15f29668bc8a8168975f6de8da926ea398619cfff5894bb71d31a777fa7ea18e"
published-version = "1"
