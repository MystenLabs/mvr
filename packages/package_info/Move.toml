[package]
name = "package_info"
edition = "2024.beta"

[dependencies]
Sui = { git = "https://github.com/MystenLabs/sui.git", subdir = "crates/sui-framework/packages/sui-framework", rev = "framework/testnet", override=true }

# goes into [dependencies] section
Codec = { git = "https://github.com/sui-potatoes/app.git", subdir = "packages/codec", rev = "codec@mainnet-v1" }


# goes into [dependencies] section
# Codec = { git = "https://github.com/sui-potatoes/app.git", subdir = "packages/codec", rev = "codec@testnet-v1" }

[addresses]
package_info="0x0"
