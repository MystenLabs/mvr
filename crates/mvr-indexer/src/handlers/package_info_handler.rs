use crate::handlers::MoveObjectProcessor;
use crate::handlers::{convert_struct_tag, OrderedDedup};
use crate::models::mainnet::mvr_metadata::package_info::PackageInfo as MainnetPkgInfo;
use crate::models::mainnet::sui::vec_map::VecMap;
use crate::models::testnet::mvr_metadata::package_info::PackageInfo as TestnetPkgInfo;
use async_trait::async_trait;
use diesel::query_dsl::methods::FilterDsl;
use diesel::upsert::excluded;
use diesel::ExpressionMethods;
use diesel_async::RunQueryDsl;
use move_types::MoveStruct;
use mvr_schema::models::PackageInfo;
use serde::de::DeserializeOwned;
use std::collections::HashMap;
use std::marker::PhantomData;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_pg_db::Connection;
use sui_sdk_types::{Address, ObjectId};
use sui_types::full_checkpoint_content::CheckpointData;
use sui_types::object::Object;

const DEFAULT_NAME_KEY: &str = "default";

pub struct PackageInfoHandler<T> {
    chain_id: String,
    phantom_data: PhantomData<T>,
}

impl<T> PackageInfoHandler<T> {
    pub fn new(chain_id: String) -> Self {
        Self {
            chain_id,
            phantom_data: Default::default(),
        }
    }
    fn package_info(
        chain_id: String,
        id: ObjectId,
        package_addr: Address,
        metadata: VecMap<String, String>,
        git_table_id: ObjectId,
        object_version: u64,
    ) -> Result<PackageInfo, anyhow::Error> {
        let metadata: HashMap<_, _> = metadata.into();
        Ok(PackageInfo {
            id: id.to_string(),
            object_version: object_version as i64,
            package_id: package_addr.to_string(),
            git_table_id: git_table_id.to_string(),
            chain_id,
            default_name: metadata.get(&DEFAULT_NAME_KEY.to_string()).cloned(),
            metadata: serde_json::to_value(metadata)?,
        })
    }
}

// impl for mainnet PackageInfo
impl MoveObjectProcessor<MainnetPkgInfo, PackageInfo> for PackageInfoHandler<MainnetPkgInfo> {
    const PROC_NAME: &'static str = "PackageInfo - Mainnet";
    fn process_move_object(
        chain_id: String,
        move_obj: MainnetPkgInfo,
        obj: &Object,
    ) -> Result<PackageInfo, anyhow::Error> {
        let MainnetPkgInfo {
            id,
            package_address,
            metadata,
            git_versioning,
            ..
        } = move_obj;
        Self::package_info(
            chain_id,
            id,
            package_address,
            metadata,
            git_versioning.id,
            obj.version().value(),
        )
    }
}

// impl for testnet PackageInfo
impl MoveObjectProcessor<TestnetPkgInfo, PackageInfo> for PackageInfoHandler<TestnetPkgInfo> {
    const PROC_NAME: &'static str = "PackageInfo - Testnet";
    fn process_move_object(
        chain_id: String,
        move_obj: TestnetPkgInfo,
        obj: &Object,
    ) -> Result<PackageInfo, anyhow::Error> {
        let TestnetPkgInfo {
            id,
            package_address,
            metadata,
            git_versioning,
            ..
        } = move_obj;
        Self::package_info(
            chain_id,
            id,
            package_address,
            metadata,
            git_versioning.id,
            obj.version().value(),
        )
    }
}

#[async_trait]
impl<T: MoveStruct + DeserializeOwned> Handler for PackageInfoHandler<T>
where
    Self: MoveObjectProcessor<T, PackageInfo>,
{
    async fn commit(values: &[Self::Value], conn: &mut Connection<'_>) -> anyhow::Result<usize> {
        use mvr_schema::schema::package_infos::columns::*;
        use mvr_schema::schema::package_infos::dsl::package_infos;

        // dedup values to avoid affecting same row twice
        let values = values.iter().cmp_dedup(
            |v| v.id.clone(),
            |v1, v2| v1.object_version.cmp(&v2.object_version),
        );

        Ok(diesel::insert_into(package_infos)
            .values(values)
            .on_conflict(id)
            .do_update()
            .set((
                package_id.eq(excluded(package_id)),
                git_table_id.eq(excluded(git_table_id)),
                default_name.eq(excluded(default_name)),
                metadata.eq(excluded(metadata)),
            ))
            .filter(object_version.lt(excluded(object_version)))
            .execute(conn)
            .await?)
    }
}

impl<T: MoveStruct + DeserializeOwned> Processor for PackageInfoHandler<T>
where
    Self: MoveObjectProcessor<T, PackageInfo>,
{
    const NAME: &'static str = Self::PROC_NAME;
    type Value = PackageInfo;

    fn process(&self, checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
        checkpoint.transactions.iter().try_fold(vec![], |result, tx| {
            tx.output_objects.iter().try_fold(result, |mut result, obj| {
                if matches!(obj.type_(), Some(t) if matches!(t.other(), Some(s) if s == &convert_struct_tag(T::struct_type())) ) {
                    if let Some(move_obj) = obj.data.try_as_move() {
                        let move_obj: T = bcs::from_bytes(move_obj.contents())?;
                        result.push(Self::process_move_object(self.chain_id.clone(), move_obj, obj)?)
                    }
                }
                Ok(result)
            })
        })
    }
}
