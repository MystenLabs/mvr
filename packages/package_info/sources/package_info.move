module package_info::package_info;

use package_info::{display::{Self, PackageDisplay}, git::GitInfo};
use std::string::String;
use sui::{
    dynamic_field as df,
    package::{Self, UpgradeCap},
    table::{Self, Table},
    vec_map::{Self, VecMap}
};

use fun df::remove as UID.remove;
use fun df::add as UID.add;

#[error]
const ECannotCreateDuringUpgrade: vector<u8> =
    b"Cannot create a package metadata object during upgrade";
#[error]
const EVersionNotFound: vector<u8> = b"Cannot remove a version that does not exist";
#[error]
const EVersionAlreadyExists: vector<u8> =
    b"Cannot override a version that already exists. Call `unset_git_versioning` first";

/// OTW to claim `Display` for this package.
public struct PACKAGE_INFO has drop {}

/// The `PackageInfo` struct holds all the metadata needed about a package.
/// This object is `key` only to make sure it's indexable at all times,
/// as it acts as the source of truth for the .move service,
/// and is guaranteed to be an owned object (key only + only transfer only
/// available)
public struct PackageInfo has key {
    id: UID,
    /// The
    display: PackageDisplay,
    // the ID of the upgrade cap
    upgrade_cap_id: ID,
    // the address of the package (no version specified, any version that got
    // attached)
    // Resolution of particular versions will occur through the RPCs versioned
    // resolvers.
    package_address: address,
    // We can hold any metadata we want for the package (up to obj size limit).
    metadata: VecMap<String, String>,
    // We can hold the git versioning here.
    git_versioning: Table<u64, GitInfo>,
}

fun init(otw: PACKAGE_INFO, ctx: &mut TxContext) {
    package::claim_and_keep(otw, ctx);
}

/// Create a new empty `PackageInfo` object for a given `upgrade_cap`.
/// Expects a `&mut UpgradeCap` for added security.
/// (All privileged upgrade cap functions require &mut)/
public fun new(cap: &mut UpgradeCap, ctx: &mut TxContext): PackageInfo {
    assert!(cap.upgrade_package().to_address() != @0x0, ECannotCreateDuringUpgrade);

    PackageInfo {
        id: object::new(ctx),
        display: display::default(b"Package".to_string()),
        package_address: cap.upgrade_package().to_address(),
        upgrade_cap_id: object::id(cap),
        metadata: vec_map::empty(),
        git_versioning: table::new(ctx),
    }
}

/// Last PTB call (or ownership change).
public fun transfer(info: PackageInfo, to: address) {
    transfer::transfer(info, to)
}

public fun set_display(info: &mut PackageInfo, mut display: PackageDisplay) {
    // we encode the label here (We get the proper SVG data in the display)
    display.encode_label(info.package_address.to_string());
    info.display = display;
}

/// Set any metadata for the NFT.
public fun set_metadata(info: &mut PackageInfo, key: String, value: String) {
    info.metadata.insert(key, value);
}

/// Unset any plain-text metadata from the NFT.
public fun unset_metadata(info: &mut PackageInfo, key: String) {
    info.metadata.remove(&key);
}

/// Allows us to set the github metadata for any given version of a package.
///
/// This is helpful for:
/// 1. Source validation services: It will work on all set versions with the
/// correct source code on those revisions.
/// 2. Development process: Easy to depend on any version of the package.
public fun set_git_versioning(info: &mut PackageInfo, version: u64, git_info: GitInfo) {
    assert!(!info.git_versioning.contains(version), EVersionAlreadyExists);
    info.git_versioning.add(version, git_info);
}

/// Allows unsetting a previously set git version.
/// This should be used to:
/// 1. Remove any invalid version.
/// 2. Override versions by unsetting + then calling "set". (updates)
public fun unset_git_versioning(info: &mut PackageInfo, version: u64): GitInfo {
    assert!(info.git_versioning.contains(version), EVersionNotFound);
    info.git_versioning.remove(version)
}

/// Allows the owner to attach any other logic / DFs to the NFT.
public fun set_custom_metadata<K: copy + store + drop, V: store>(
    info: &mut PackageInfo,
    key: K,
    value: V,
) {
    info.id.add(key, value)
}

/// Allows removing any custom metadata from the NFT.
public fun remove_custom_metadata<K: copy + store + drop, V: store>(
    info: &mut PackageInfo,
    key: K,
): V {
    info.id.remove(key)
}

/// === Getters ===
public fun id(info: &PackageInfo): ID {
    info.id.to_inner()
}

public fun package_address(info: &PackageInfo): address {
    info.package_address
}

public fun upgrade_cap_id(info: &PackageInfo): ID {
    info.upgrade_cap_id
}
