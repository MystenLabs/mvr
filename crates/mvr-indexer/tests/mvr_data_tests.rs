use chrono::NaiveDateTime;
use fastcrypto::hash::{HashFunction, Sha256};
use insta::assert_json_snapshot;
use mvr_indexer::handlers::git_info_handler::GitInfoHandler;
use mvr_indexer::handlers::git_info_handler::MainnetGitInfo;
use mvr_indexer::handlers::name_record_handler::NameRecordHandler;
use mvr_indexer::handlers::package_handler::PackageHandler;
use mvr_indexer::handlers::package_info_handler::PackageInfoHandler;
use mvr_indexer::models::mainnet::mvr_metadata::package_info::PackageInfo;
use mvr_indexer::MAINNET_CHAIN_ID;
use mvr_schema::MIGRATIONS;
use serde_json::Value;
use sqlx::{Column, PgPool, Row, ValueRef};
use std::fs;
use std::path::Path;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_indexer_alt_framework::store::Store;
use sui_pg_db::temp::TempDb;
use sui_pg_db::Connection;
use sui_pg_db::Db;
use sui_pg_db::DbArgs;
use sui_storage::blob::Blob;
use sui_types::full_checkpoint_content::CheckpointData;

#[tokio::test]
async fn git_info_write_test() -> Result<(), anyhow::Error> {
    let handler = GitInfoHandler::<MainnetGitInfo>::new(MAINNET_CHAIN_ID.into());
    data_test("git_infos_write", handler, ["git_infos"]).await?;
    Ok(())
}

#[tokio::test]
async fn package_write_test() -> Result<(), anyhow::Error> {
    let handler = PackageHandler::<true>;
    data_test(
        "packages_write",
        handler,
        ["packages", "package_dependencies"],
    )
    .await?;
    Ok(())
}

#[tokio::test]
async fn package_infos_write_test() -> Result<(), anyhow::Error> {
    let handler = PackageInfoHandler::<PackageInfo>::new(MAINNET_CHAIN_ID.into());
    data_test("package_infos_write", handler, ["package_infos"]).await?;
    Ok(())
}

#[tokio::test]
async fn name_records_write_test() -> Result<(), anyhow::Error> {
    let handler = NameRecordHandler::new();
    data_test("name_records_write", handler, ["name_records"]).await?;
    Ok(())
}

#[tokio::test]
async fn immediate_dependency_test() -> Result<(), anyhow::Error> {
    let handler = PackageHandler::<true>;
    data_test(
        "immediate_dependency",
        handler,
        ["packages", "package_dependencies"],
    )
    .await?;
    Ok(())
}

async fn data_test<H, I>(
    test_name: &str,
    handler: H,
    tables_to_check: I,
) -> Result<(), anyhow::Error>
where
    I: IntoIterator<Item = &'static str>,
    H: Handler + Processor,
    for<'a> H::Store: Store<Connection<'a> = Connection<'a>>,
{
    // Set up the temporary database
    let temp_db = TempDb::new()?;
    let url = temp_db.database().url();
    let db = Arc::new(Db::for_write(url.clone(), DbArgs::default()).await?);
    db.run_migrations(MIGRATIONS).await?;
    let mut conn = db.connect().await?;

    // Test setup based on provided test_name
    let test_path = Path::new("tests/checkpoints").join(test_name);
    let checkpoints = get_checkpoints_in_folder(&test_path)?;

    // Run pipeline for each checkpoint
    for checkpoint in checkpoints {
        run_pipeline(&handler, &checkpoint, &mut conn).await?;
    }

    // Check results by comparing database tables with snapshots
    for table in tables_to_check {
        let rows = read_table(&table, &url.to_string()).await?;
        assert_json_snapshot!(format!("{test_name}__{table}"), rows);
    }
    Ok(())
}

async fn run_pipeline<'c, T: Handler + Processor, P: AsRef<Path>>(
    handler: &T,
    path: P,
    conn: &mut Connection<'c>,
) -> Result<(), anyhow::Error>
where
    T::Store: Store<Connection<'c> = Connection<'c>>,
{
    let bytes = fs::read(path)?;
    let cp = Blob::from_bytes::<CheckpointData>(&bytes)?;
    let result = handler.process(&Arc::new(cp))?;
    T::commit(&result, conn).await?;
    Ok(())
}

/// Read the entire table from database as json value.
/// note: bytea values will be hashed to reduce output size.
async fn read_table(table_name: &str, db_url: &str) -> Result<Vec<Value>, anyhow::Error> {
    let pool = PgPool::connect(db_url).await?;
    let rows = sqlx::query(&format!("SELECT * FROM {table_name}"))
        .fetch_all(&pool)
        .await?;

    // To json
    Ok(rows
        .iter()
        .map(|row| {
            let mut obj = serde_json::Map::new();

            for column in row.columns() {
                let column_name = column.name();

                let value = if let Ok(v) = row.try_get::<String, _>(column_name) {
                    Value::String(v)
                } else if let Ok(v) = row.try_get::<i32, _>(column_name) {
                    Value::Number(v.into())
                } else if let Ok(v) = row.try_get::<i64, _>(column_name) {
                    Value::Number(v.into())
                } else if let Ok(v) = row.try_get::<bool, _>(column_name) {
                    Value::Bool(v)
                } else if let Ok(v) = row.try_get::<Value, _>(column_name) {
                    v
                } else if let Ok(v) = row.try_get::<Vec<u8>, _>(column_name) {
                    // hash bytea contents
                    let mut hash_function = Sha256::default();
                    hash_function.update(v);
                    let digest2 = hash_function.finalize();
                    Value::String(digest2.to_string())
                } else if let Ok(v) = row.try_get::<NaiveDateTime, _>(column_name) {
                    Value::String(v.to_string())
                } else if let Ok(true) = row.try_get_raw(column_name).map(|v| v.is_null()) {
                    Value::Null
                } else {
                    panic!(
                        "Cannot parse DB value to json, type: {:?}, column: {column_name}",
                        row.try_get_raw(column_name)
                            .map(|v| v.type_info().to_string())
                    )
                };
                obj.insert(column_name.to_string(), value);
            }

            Value::Object(obj)
        })
        .collect())
}

fn get_checkpoints_in_folder(folder: &Path) -> Result<Vec<String>, anyhow::Error> {
    let mut files = Vec::new();

    // Read the directory
    for entry in fs::read_dir(folder)? {
        let entry = entry?;
        let path = entry.path();

        // Check if it's a file and ends with ".chk"
        if path.is_file() && path.extension().and_then(|s| s.to_str()) == Some("chk") {
            files.push(path.display().to_string());
        }
    }

    Ok(files)
}
