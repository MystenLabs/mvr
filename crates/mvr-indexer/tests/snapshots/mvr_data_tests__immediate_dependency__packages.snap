---
source: crates/mvr-indexer/tests/mvr_data_tests.rs
expression: rows
---
[
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "original_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "package_version": 1,
    "chain_id": "35834a8a",
    "move_package": "4JTuCBbZpDPkxaHo5qOk7uxYqCshHjIxTmIxxFF3qt4=",
    "tx_hash": "C2L7ChM9DQcAhugbfNQjNL7aVQEZcCdSAyxA7muSqon7",
    "sender": "0xeb8a8770e34c76d62c9e692c894adb1fdd775a71bc78ed54c6dca1a368e779f5",
    "timestamp": "2024-12-19 06:16:52.373"
  }
]
